File: algorithms/trees/binary-search-tree/csharp/BinarySearchTree.cs
using System;
using System.Collections.Generic;

public class BinarySearchTree
{
    private class Node
    {
        public int Key;
        public Node Left;
        public Node Right;

        public Node(int key)
        {
            Key = key;
        }
    }

    private static Node Insert(Node root, int key)
    {
        if (root == null)
        {
            return new Node(key);
        }
        if (key <= root.Key)
        {
            root.Left = Insert(root.Left, key);
        }
        else
        {
            root.Right = Insert(root.Right, key);
        }
        return root;
    }

    private static void Inorder(Node root, List<int> result)
    {
        if (root == null) return;
        Inorder(root.Left, result);
        result.Add(root.Key);
        Inorder(root.Right, result);
    }

    public static int[] BstInorder(int[] arr)
    {
        Node root = null;
        foreach (int key in arr)
        {
            root = Insert(root, key);
        }

        List<int> result = new List<int>();
        Inorder(root, result);
        return result.ToArray();
    }
}

File: algorithms/trees/binary-search-tree/scala/BinarySearchTree.scala
object BinarySearchTree {

  private class Node(val key: Int) {
    var left: Node = _
    var right: Node = _
  }

  private def insert(root: Node, key: Int): Node = {
    if (root == null) {
      return new Node(key)
    }
    if (key <= root.key) {
      root.left = insert(root.left, key)
    } else {
      root.right = insert(root.right, key)
    }
    root
  }

  private def inorder(root: Node, result: scala.collection.mutable.ListBuffer[Int]): Unit = {
    if (root == null) return
    inorder(root.left, result)
    result += root.key
    inorder(root.right, result)
  }

  def bstInorder(arr: Array[Int]): Array[Int] = {
    var root: Node = null
    for (key <- arr) {
      root = insert(root, key)
    }

    val result = scala.collection.mutable.ListBuffer[Int]()
    inorder(root, result)
    result.toArray
  }
}
